packages/mui-base/src/utils/useCompound.ts

```diff
@@ -0,0 +1,84 @@
+import * as React from 'react';
+
+export type CompoundComponentContextValue<Key, Subitem> = {
+  registerItem: (id: Key, item: Subitem) => void;
+  unregisterItem: (id: Key) => void;
+  getItemIndex: (id: Key) => number;
+  totalSubitems: number;
+};
+
+export const CompoundComponentContext = React.createContext<CompoundComponentContextValue<
+  any,
+  any
+> | null>(null);
+
+CompoundComponentContext.displayName = 'CompoundComponentContext';
+
+export interface UseCompoundParentParameters<Key, Subitem> {
+  onRegisterItem?: (id: Key, item: Subitem) => void;
+  onUnregisterItem?: (id: Key) => void;
+}
+
+export default function useCompoundParent<Key, Subitem>() {
+  const [subitems, setSubitems] = React.useState(new Map<Key, Subitem>());
+
+  const registerItem = React.useCallback(function registerItem(id: Key, item: Subitem) {
+    setSubitems((previousState) => {
+      const newState = new Map(previousState);
+      newState.set(id, item);
+      return newState;
+    });
+  }, []);
+
+  const unregisterItem = React.useCallback(function unregisterItem(id: Key) {
+    setSubitems((previousState) => {
+      const newState = new Map(previousState);
+      newState.delete(id);
+      return newState;
+    });
+  }, []);
+
+  const getItemIndex = React.useCallback(
+    function getItemIndex(id: Key) {
+      return Array.from(subitems.keys()).indexOf(id);
```

Comment:
I needed to check if the keys are sorted in insertion order :D I thought they are in lexicological order, but I guess that comes from a different language.